Xamarin.PropertyEditing.Windows/TreeViewItemEx.cs

```diff
@@ -0,0 +1,264 @@
+			get { return (Thickness)GetValue (ChildMarginProperty); }
+			set { SetValue (ChildMarginProperty, value); }
+		}
+
+		private static readonly DependencyPropertyKey IsMouseOverItemKey = DependencyProperty.RegisterReadOnly ("IsMouseOverItem", typeof(bool), typeof(TreeViewItemEx), new PropertyMetadata());
+		public static readonly DependencyProperty IsMouseOverItemProperty = IsMouseOverItemKey.DependencyProperty;
+
+		public bool IsMouseOverItem
+		{
+			get { return (bool)GetValue (IsMouseOverItemProperty); }
+			private set { SetValue (IsMouseOverItemKey, value); }
+		}
+
+		public static readonly DependencyProperty IsSelectableProperty = DependencyProperty.Register ("IsSelectable", typeof(bool), typeof(TreeViewItemEx), new PropertyMetadata (true));
+
+		public bool IsSelectable
+		{
+			get { return (bool)GetValue (IsSelectableProperty); }
+			set { SetValue (IsSelectableProperty, value); }
+		}
+
+		public override void OnApplyTemplate ()
+		{
+			base.OnApplyTemplate ();
+
+			int level = 0;
+			var parent = VisualTreeHelper.GetParent (this);
+			while (!(parent is TreeView)) {
+				parent = VisualTreeHelper.GetParent (parent);
+
+				if (parent is TreeViewItem)
+					level++;
+			}
+
+			this.topParent = parent as TreeViewEx;
+
+			IndentLevel = level;
+		}
+
+		protected override void OnMouseDown (MouseButtonEventArgs e)
+		{
+			if (IsSelectable) {
+				if (Keyboard.IsKeyDown (Key.LeftCtrl) || Keyboard.IsKeyDown (Key.RightCtrl)) {
+					IsSelected = !IsSelected;
+					e.Handled = true;
+				}
+			}
+
+			base.OnMouseDown (e);
+		}
+
+		protected override void OnKeyDown (KeyEventArgs e)
+		{
+			if (e.Key == Key.Up || e.Key == Key.Down) {
+				var next = GetNextElement (e.Key == Key.Up) as ItemsControl;
+				if (!(next is TreeView)) {
+					if (next.HasItems) {
+						var tree = GetParentTree();
+						tree.SelectedTreeItem = null;
+					}
+
+					next.Focus();
+					e.Handled = true;
+				}
+			}
+
+			base.OnKeyDown (e);
+		}
+
+		protected override void OnSelected (RoutedEventArgs e)
+		{
+			TreeViewEx parent = GetParentTree();
+			if (!IsSelectable) {
+				IsSelected = false;
+				e.Handled = true;
+
+				parent.SelectedTreeItem = parent.SelectedTreeItem;
+			} else if (parent.SelectedTreeItem != this) {
+				parent.SelectedTreeItem = this;
+			}
+		}
+
+		protected override DependencyObject GetContainerForItemOverride ()
+		{
+			return new TreeViewItemEx();
+		}
+
+		protected override bool IsItemItsOwnContainerOverride (object item)
+		{
+			return item is TreeViewItemEx;
+		}
+
+		protected override void OnMouseEnter (MouseEventArgs e)
+		{
+			base.OnMouseEnter (e);
+
+			CheckPosition (e);
+		}
+
+		protected override void OnMouseMove (MouseEventArgs e)
+		{
+			base.OnMouseMove (e);
+
+			CheckPosition (e);
+		}
+
+		protected override void OnMouseLeave (MouseEventArgs e)
+		{
+			base.OnMouseLeave (e);
+			IsMouseOverItem = false;
+		}
+
+		private TreeViewEx topParent;
+
+		private UIElement GetNextElement (bool previous)
```

Comment:
I got confused for a second by this until I looked up usage: next and previous are contradicting each other, so maybe rename that parameter to something like `searchUpwards`?

Reply:
We should probably have some tests on that method.

Reply:
We don't really have a UI testing infrastructure currently.

Reply:
I meant unit test that one method.